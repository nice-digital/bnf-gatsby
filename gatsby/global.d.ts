// Let TypeScript know we're importing SCSS (and SCSS modules)
// Avoids "Can't import CSS/SCSS modules. TypeScript says “Cannot Find Module”"
// See https://stackoverflow.com/a/41946697/486434
declare module "*.module.scss" {
	const content: { [className: string]: string };
	export default content;
}

// So that we can import JSON files in TypeScript.
declare module "*.json" {
	const value: unknown;

	export default value;
}
