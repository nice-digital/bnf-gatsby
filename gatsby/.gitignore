node_modules/
.cache/
public

# dotenv environment variables file
.env
.env.*
