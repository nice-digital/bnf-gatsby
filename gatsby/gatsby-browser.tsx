import {
	ShouldUpdateScrollArgs,
	type RouteUpdateArgs,
	type WrapPageElementBrowserArgs,
} from "gatsby";
import { type ReactElement } from "react";

import { Layout } from "@/components/Layout/Layout";

// Gatsby hook for when the route has changed on the client side
// See https://www.gatsbyjs.org/docs/browser-apis/#onRouteUpdate
export const onRouteUpdate = ({
	prevLocation,
	location,
}: RouteUpdateArgs): void => {
	if (prevLocation) {
		// Cast is needed because of https://github.com/gatsbyjs/gatsby/issues/29124
		const prevPath = (prevLocation as Location).pathname,
			path = location.pathname;

		// Push our own event to the dataLayer on page change rather than using the
		// 'gatsby-route-change' event built into gatsby-plugin-google-tagmanager.
		// Because gatsby-route-change is pushed on initial page load as well as route change,
		// AND because it doesn't use requestAnimationFrame to delay until the page title has been updated

		const sendPageView = () => {
			window.dataLayer.push({
				location: location.href,
				referrer: prevLocation.href,
			});
			// Don't consider hash changes to be a page view - pageviews only happy when the path changes
			if (prevPath != path) window.dataLayer.push({ event: "pageview" });
		};

		// Delay before push to the data layer, to make sure the page title has been updated
		// See https://github.com/gatsbyjs/gatsby/pull/10917/files#diff-bf0d94c8bf47d5c1687e342c2dba1e00R12-R13
		if ("requestAnimationFrame" in window) {
			window.requestAnimationFrame(() => {
				window.requestAnimationFrame(sendPageView);
			});
		} else {
			// simulate 2 rAF calls
			setTimeout(sendPageView, 32);
		}
	} else {
		window.dataLayer.push({
			location: location.href,
			referrer: document.referrer,
		});
	}
	// Default return statement when prevLocation is falsy
	return;
};

export const wrapPageElement = ({
	element,
	props,
}: WrapPageElementBrowserArgs): ReactElement => (
	<Layout {...props}>{element}</Layout>
);

/**
 * Gatsby hook for overriding scroll position
 * See https://www.gatsbyjs.org/docs/browser-apis/#shouldUpdateScroll
 */
export const shouldUpdateScroll = ({
	routerProps: { location },
	prevRouterProps,
	getSavedScrollPosition,
}: ShouldUpdateScrollArgs): boolean | string | [number, number] => {
	if (
		// If there's no previous route props we're coming from an external site, which means
		// we want to scroll to the hash (if there is one), and _not_ a saved scroll position
		!prevRouterProps ||
		// Or we're linking to as hash within the same page
		(prevRouterProps.location.pathname === location.pathname && location.hash)
	) {
		// Provide our own scroll to hash to avoid Gatsby using a stored scroll position
		requestAnimationFrame(() => {
			requestAnimationFrame(() => {
				const targetElement = location.hash
					? document.querySelector(location.hash)
					: null;

				if (targetElement) {
					targetElement.setAttribute("tabIndex", "-1");
					(targetElement as HTMLElement).focus();
					targetElement.scrollIntoView();
				}
			});
		});
		return false;
	}

	const savedScrollY = (getSavedScrollPosition(location)?.[1] || 0) as number;
	if (savedScrollY > 0) {
		window.scrollTo(0, savedScrollY);
		return false;
	}

	if (location.hash) {
		requestAnimationFrame(() => {
			requestAnimationFrame(() => {
				const targetElement = document.querySelector(location.hash);

				// Default to Gatsby's default behaviour if the element doesn't exist
				if (!targetElement) return true;

				// Provide our own scroll to hash to avoid Gatsby using a stored scroll position
				targetElement.setAttribute("tabIndex", "-1");
				(targetElement as HTMLElement).focus();
				targetElement.scrollIntoView();
				return false;
			});
		});
		return false;
	}

	requestAnimationFrame(() => {
		requestAnimationFrame(() => {
			const contentStartElement = document.getElementById("content-start");

			if (!contentStartElement) return true;

			contentStartElement.setAttribute("tabIndex", "-1");

			//scrolling to 0,0 to preserve existing BNF navigation behaviour
			window.scrollTo(0, 0);
			return false;
		});
	});
	// Default to scrolling to the content start element as the standard navigation behaviour

	return true;
};
