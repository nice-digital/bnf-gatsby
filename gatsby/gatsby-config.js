require("dotenv").config({
	path: `.env.${process.env.NODE_ENV}`,
});

const isBNF = process.env.GATSBY_SITE === "bnf";

if (!process.env.GATSBY_SEARCH_URL)
	throw new Error(
		"Env var GATSBY_SEARCH_URL isn't set. Did you forget to add it?"
	);

module.exports = {
	jsxRuntime: "automatic",
	siteMetadata: {
		isBNF,
		siteUrl: `https://${isBNF ? "bnf" : "bnfc"}.nice.org.uk`,
		siteTitleShort: isBNF ? "BNF" : "BNFC",
		siteTitleLong:
			"British National Formulary" + (isBNF ? "" : " for Children"),
	},
	plugins: [
		{
			resolve: `gatsby-source-bnf`,
			options: {
				feedURL: process.env.FEED_URL,
				userKey: process.env.FEED_USER_KEY,
				site: isBNF ? "bnf" : "bnfc",
			},
		},
		// Avoid errors like "ModuleNotFoundError: Module not found: Error: Can't resolve '@/components/Layout/Layout'" when using custom paths in tsconfig.json
		`gatsby-plugin-tsconfig-paths`,
		{
			resolve: `gatsby-plugin-sass`,
			options: {
				// Make CSS modules play nicely with TypeScript, see also global.d.ts for *.scss TS def
				cssLoaderOptions: {
					esModule: false,
					modules: {
						namedExport: false,
					},
				},
			},
		},
		"gatsby-plugin-catch-links",
		"gatsby-plugin-react-helmet",
		"gatsby-plugin-remove-generator",
		{
			resolve: `gatsby-plugin-sitemap`,
			options: {
				excludes: [`/search/`],
			},
		},
		{
			resolve: "gatsby-plugin-google-tagmanager",
			options: {
				id: "GTM-5H5L9GK",
				includeInDevelopment: true,
			},
		},
		{
			resolve: "gatsby-plugin-manifest",
			options: {
				icon: "src/images/icon.png",
			},
		},
	],
};
