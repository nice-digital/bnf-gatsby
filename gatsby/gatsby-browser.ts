import { RouteUpdateArgs } from "gatsby";

// Gatsby hook for when the route has changed on the client side
// See https://www.gatsbyjs.org/docs/browser-apis/#onRouteUpdate
export const onRouteUpdate = ({
	prevLocation,
	location,
}: RouteUpdateArgs): void => {
	if (prevLocation) {
		// Cast is needed because of https://github.com/gatsbyjs/gatsby/issues/29124
		const prevPath = (prevLocation as Location).pathname,
			path = location.pathname;

		// Push our own event to the dataLayer on page change rather than using the
		// 'gatsby-route-change' event built into gatsby-plugin-google-tagmanager.
		// Because gatsby-route-change is pushed on initial page load as well as route change,
		// AND because it doesn't use requestAnimationFrame to delay until the page title has been updated

		const sendPageView = () => {
			window.dataLayer.push({
				location: location.href,
				referrer: prevLocation.href,
			});
			// Don't consider hash changes to be a page view - pageviews only happy when the path changes
			if (prevPath != path) window.dataLayer.push({ event: "pageview" });
		};

		// Delay before push to the data layer, to make sure the page title has been updated
		// See https://github.com/gatsbyjs/gatsby/pull/10917/files#diff-bf0d94c8bf47d5c1687e342c2dba1e00R12-R13
		if ("requestAnimationFrame" in window) {
			window.requestAnimationFrame(() => {
				window.requestAnimationFrame(sendPageView);
			});
		} else {
			// simulate 2 rAF calls
			setTimeout(sendPageView, 32);
		}
	} else {
		window.dataLayer.push({
			location: location.href,
			referrer: document.referrer,
		});
	}
};
