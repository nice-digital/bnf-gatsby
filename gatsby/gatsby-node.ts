import { CreateSchemaCustomizationArgs, CreateWebpackConfigArgs } from "gatsby";

import { initialFieldExtension } from "./src/field-extensions/initial";

/**
 * Gatsby hook for customizing the schema.
 * See https://www.gatsbyjs.org/docs/schema-customization/
 */
export const createSchemaCustomization = ({
	actions: { createFieldExtension, createTypes },
}: CreateSchemaCustomizationArgs): void => {
	createFieldExtension(initialFieldExtension);

	// Add initials for types used in A-Z lists, as it's easier for grouping
	// We don't add the initial field in the source plugin because grouping is a presentation concern
	const typeDefs = `
	  type BnfDrug implements Node {
		"The lowercase first letter of the title, used for grouping"
		initial: String! @initial(field: "title")
	  }
	  type BnfTreatmentSummary implements Node {
		"The lowercase first letter of the title, used for grouping"
		initial: String! @initial(field: "title")
	  }
	  type BnfInteractant implements Node {
		"The lowercase first letter of the title, used for grouping"
		initial: String! @initial(field: "title")
	  }
	`;
	createTypes(typeDefs);
};

/**
 * Gatsby hook for overriding webpack config
 * See https://www.gatsbyjs.com/docs/how-to/custom-configuration/add-custom-webpack-config/
 */
export const onCreateWebpackConfig = ({
	stage,
	actions,
	getConfig,
	plugins,
}: CreateWebpackConfigArgs): void => {
	// Disable order warnings from mini CSS plugin
	// See https://github.com/gatsbyjs/gatsby/discussions/30169#discussioncomment-621285
	if (stage === "build-javascript" || stage === "develop") {
		const config = getConfig();

		const miniCss = config.plugins.find(
			(plugin: () => void) => plugin.constructor.name === "MiniCssExtractPlugin"
		);

		if (miniCss) {
			miniCss.options.ignoreOrder = true;
		}

		actions.replaceWebpackConfig(config);

		actions.setWebpackConfig({
			plugins: [plugins.provide({ process: "process/browser" })],
		});
	}
};
