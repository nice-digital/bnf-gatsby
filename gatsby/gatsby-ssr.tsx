import { PreRenderHTMLArgs } from "gatsby";
import React from "react";

export const onPreRenderHTML = ({
	getHeadComponents,
	replaceHeadComponents,
}: PreRenderHTMLArgs): void => {
	const components = getHeadComponents();

	components.push(
		<script
			src={process.env.GATSBY_COOKIE_BANNER_URL}
			type="text/javascript"
			key="cookie-banner"
			async
		></script>
	);

	replaceHeadComponents(components);
};
