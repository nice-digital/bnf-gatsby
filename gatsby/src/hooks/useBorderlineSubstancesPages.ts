import { useStaticQuery, graphql } from "gatsby";

import { SID, PHPID } from "@nice-digital/gatsby-source-bnf";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface BorderlineSubstancesQueryResult {
	allBnfBorderlineSubstancesTaxonomyRoot: {
		pages: {
			slug: string;
			title: string;
		}[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/borderline-substances/${slug}/`,
});

/**
 * Gets the lists of pages for the borderline substances section
 *
 * @returns The consolidated list of borderline substances pages
 */
export const useBorderlineSubstancesPages = (): MenuPageLink[] => {
	const { allBnfBorderlineSubstancesTaxonomyRoot } =
		useStaticQuery<BorderlineSubstancesQueryResult>(
			graphql`
				{
					allBnfBorderlineSubstancesTaxonomyRoot {
						pages: nodes {
							slug
							title
						}
					}
				}
			`
		);

	return allBnfBorderlineSubstancesTaxonomyRoot.pages.map(slugToHref);
};
