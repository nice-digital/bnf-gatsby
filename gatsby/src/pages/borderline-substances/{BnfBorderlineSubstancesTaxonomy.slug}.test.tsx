import { render, waitFor, screen, within } from "@testing-library/react";
import { useStaticQuery } from "gatsby";

import { mockBorderlineSubstancesPagesQueryData } from "@/hooks/useBorderlineSubstancesPages.test";

import BorderlineSubstancesSectionPage, {
	query,
	type BorderlineSubstancesSectionPageProps,
} from "./{BnfBorderlineSubstancesTaxonomy.slug}";

const isRootTwoLevelProps: BorderlineSubstancesSectionPageProps = {
	data: {
		bnfBorderlineSubstancesTaxonomy: {
			slug: "root-taxonomy",
			title: "Root taxonomy",
			rootTaxonomy: {
				slug: "root-taxonomy",
				title: "Root taxonomy",
				childTaxonomies: [
					{
						title: "Taxonomy 1",
						slug: "taxonomy-1",
						childTaxonomies: [
							{ title: "Sub taxonomy 1", slug: "sub-taxonomy-1" },
							{ title: "Sub taxonomy 2", slug: "sub-taxonomy-2" },
						],
					},
					{
						title: "Taxonomy 2",
						slug: "taxonomy-2",
					},
				],
			},
		},
	},
};

const isRootSingleLevelProps: BorderlineSubstancesSectionPageProps = {
	data: {
		bnfBorderlineSubstancesTaxonomy: {
			slug: "root-taxonomy",
			title: "Root taxonomy",
			rootTaxonomy: {
				slug: "root-taxonomy",
				title: "Root taxonomy",
				childTaxonomies: [
					{
						title: "Taxonomy 1",
						slug: "taxonomy-1",
					},
					{
						title: "Taxonomy 2",
						slug: "taxonomy-2",
					},
				],
			},
		},
	},
};

const isNotRootProps: BorderlineSubstancesSectionPageProps = {
	data: {
		bnfBorderlineSubstancesTaxonomy: {
			slug: "taxonomy-1",
			title: "Taxonomy 1",
			rootTaxonomy: {
				slug: "root-taxonomy",
				title: "Root taxonomy",
				childTaxonomies: [
					{
						title: "Taxonomy 1",
						slug: "taxonomy-1",
						childTaxonomies: [
							{ title: "Sub taxonomy 1", slug: "sub-taxonomy-1" },
							{ title: "Sub taxonomy 2", slug: "sub-taxonomy-2" },
						],
					},
					{
						title: "Taxonomy 2",
						slug: "taxonomy-2",
						childTaxonomies: [],
					},
				],
			},
			substances: [
				{
					title: "Fresubin® 1500 Complete",
					presentations: [
						{
							acbs: [
								"<p><b>Standard ACBS indications:</b> Disease-related malnutrition, intractable malabsorption, pre-operative preparation of malnourished patients, dysphagia, proven inflammatory bowel disease, following total gastrectomy, short-bowel syndrome, bowel fistula except bowel fistula and pre-operative preparation of malnourished patients.</p>",
							],
							carbohydrateGrams: "13",
							energyKCal: "100",
							energyKj: "420",
							fatGrams: "3.4",
							fibreGrams: "1.5",
							fibreConstituents: null,
							formulation: "Liquid (tube feed) per 100 mL",
							presentationNote: null,
							proteinGrams: "3.8",
							fatConstituents: [],
							carbohydrateConstituents: ["(sugars 0.9 g)"],
							proteinConstituents: ["milk, soya proteins"],
							borderlineSubstancePreps: [
								{
									manufacturer: "Fresenius Kabi Ltd",
									name: "Fresubin 1500 Complete liquid",
									packs: [
										{
											unit: "litre",
											size: "1.5",
											nhsIndicativePrice: "£15.32",
											legalCategory: "Not Applicable",
											hospitalOnly: false,
											drugTariffPrice: null,
											drugTariff: null,
											colour: null,
											amppId: "16052511000001104",
										},
									],
									ampId: "16052411000001103",
									blackTriangle: false,
									controlledDrugSchedule: null,
									activeIngredients: [],
									sugarFree: null,
								},
							],
							rxAdvice: null,
							specialCharacteristics: [
								"Contains fish oil, residual lactose, soya. Gluten-free",
							],
						},
					],
					introductionNote:
						"<p>Not suitable for use in child under 3 years; not recommended for child under 6 years</p>",
					id: "PHP103613",
				},
				{
					title: "Fresubin® Original Fibre",
					presentations: [
						{
							acbs: [
								"<p><b>Standard ACBS indications:</b> Disease-related malnutrition, intractable malabsorption, pre-operative preparation of malnourished patients, dysphagia, proven inflammatory bowel disease, following total gastrectomy, short-bowel syndrome, bowel fistula except bowel fistula and pre-operative preparation of malnourished patients.</p>",
							],
							carbohydrateGrams: "13.0",
							energyKCal: "100",
							energyKj: "420",
							fatGrams: "3.4",
							fibreConstituents: null,
							fibreGrams: "1.5",
							formulation: "Liquid (tube feed) per 100 mL",
							presentationNote: null,
							proteinGrams: "3.8",
							fatConstituents: [],
							carbohydrateConstituents: ["(sugars 0.9 g)"],
							proteinConstituents: ["milk, soya proteins"],
							borderlineSubstancePreps: [
								{
									manufacturer: "Fresenius Kabi Ltd",
									name: "Fresubin Original Fibre liquid",
									packs: [
										{
											unit: "ml",
											size: "500",
											nhsIndicativePrice: "£5.97",
											legalCategory: "Not Applicable",
											hospitalOnly: false,
											drugTariffPrice: null,
											drugTariff: null,
											colour: null,
											amppId: "1690511000001101",
										},
										{
											unit: "ml",
											size: "1000",
											nhsIndicativePrice: "£11.26",
											legalCategory: "Not Applicable",
											hospitalOnly: false,
											drugTariffPrice: null,
											drugTariff: null,
											colour: null,
											amppId: "1690611000001102",
										},
									],
									ampId: "871011000001101",
									blackTriangle: false,
									controlledDrugSchedule: null,
									activeIngredients: [],
									sugarFree: null,
								},
							],
							rxAdvice: null,
							specialCharacteristics: [
								"Contains fish oil, residual lactose, soya. Gluten-free",
							],
						},
					],
					introductionNote:
						"<p>Not suitable for use in child under 3 years; not recommended for child under 6 years</p>",
					id: "PHP103612",
				},
			],
			childTaxonomies: [
				{
					title: "Sub taxonomy 1",
					substances: [
						{
							title: "Fresubin® Soya Fibre",
							introductionNote:
								"<p>Not suitable for use in child under 3 years; not recommended for child under 6 years</p>",
							id: "PHP103619",
							presentations: [
								{
									acbs: [
										"<p><b>Standard ACBS indications:</b> Disease-related malnutrition, intractable malabsorption, pre-operative preparation of malnourished patients, dysphagia, proven inflammatory bowel disease, following total gastrectomy, short-bowel syndrome, bowel fistula except bowel fistula and pre-operative preparation of malnourished patients.</p>",
									],
									carbohydrateGrams: "13.0",
									energyKCal: "100",
									energyKj: "420",
									fatGrams: "3.4",
									fibreConstituents: null,
									fibreGrams: "1.5",
									formulation: "Liquid (tube feed) per 100 mL",
									presentationNote: null,
									proteinGrams: "3.8",
									fatConstituents: [],
									carbohydrateConstituents: ["(sugars 0.9 g)"],
									proteinConstituents: ["milk, soya proteins"],
									borderlineSubstancePreps: [
										{
											manufacturer: "Fresenius Kabi Ltd",
											name: "Fresubin Original Fibre liquid",
											packs: [
												{
													unit: "ml",
													size: "500",
													nhsIndicativePrice: "£5.97",
													legalCategory: "Not Applicable",
													hospitalOnly: false,
													drugTariffPrice: null,
													drugTariff: null,
													colour: null,
													amppId: "1690511000001101",
												},
												{
													unit: "ml",
													size: "1000",
													nhsIndicativePrice: "£11.26",
													legalCategory: "Not Applicable",
													hospitalOnly: false,
													drugTariffPrice: null,
													drugTariff: null,
													colour: null,
													amppId: "1690611000001102",
												},
											],
											ampId: "871011000001101",
											blackTriangle: false,
											controlledDrugSchedule: null,
											activeIngredients: [],
											sugarFree: null,
										},
									],
									rxAdvice: null,
									specialCharacteristics: [
										"Contains fish oil, residual lactose, soya. Gluten-free",
									],
								},
							],
						},
					],
					childTaxonomies: [],
				},
				{ title: "Sub taxonomy 2", substances: [], childTaxonomies: [] },
			],
		},
	},
};

describe("BorderlineSubstancesSectionPage", () => {
	beforeEach(() => {
		(useStaticQuery as jest.Mock).mockReturnValue(
			mockBorderlineSubstancesPagesQueryData
		);
	});

	it("should match snapshot for graphql query", () => {
		expect(query).toMatchSnapshot();
	});

	describe("SEO", () => {
		it("should set page title from borderline substance taxonomy title", async () => {
			render(<BorderlineSubstancesSectionPage {...isRootTwoLevelProps} />);

			await waitFor(() => {
				expect(document.title).toStartWith("Root taxonomy |");
			});
		});
	});

	describe("Page header", () => {
		it("should add content start skip link target id to page header", () => {
			render(<BorderlineSubstancesSectionPage {...isRootTwoLevelProps} />);
			const heading1 = screen.getByRole("heading", {
				level: 1,
			});
			// eslint-disable-next-line testing-library/no-node-access
			expect(heading1.parentElement).toHaveAttribute("id", "content-start");
		});

		it("should render heading 1 with current page title", () => {
			render(<BorderlineSubstancesSectionPage {...isRootTwoLevelProps} />);
			const heading1 = screen.getByRole("heading", {
				level: 1,
			});
			expect(heading1).toHaveTextContent("Root taxonomy");
		});
	});

	describe("Breadcrumbs", () => {
		it.each([
			["NICE", "https://www.nice.org.uk/"],
			["BNF", "/"],
			["Borderline substances", "/borderline-substances/"],
		])(
			"should render default '(%s)' breadcrumb",
			(breadcrumbText, expectedHref) => {
				render(<BorderlineSubstancesSectionPage {...isRootTwoLevelProps} />);

				const breadcrumbNav = screen.getByRole("navigation", {
					name: "Breadcrumbs",
				});
				const breadcrumb = within(breadcrumbNav).getByRole("link", {
					name: breadcrumbText,
				});

				expect(breadcrumb).toHaveAttribute("href", expectedHref);
			}
		);

		it("should render current page breadcrumb without link", () => {
			render(<BorderlineSubstancesSectionPage {...isRootTwoLevelProps} />);

			const breadcrumbNav = screen.getByRole("navigation", {
				name: "Breadcrumbs",
			});
			const currentPageCrumb = within(breadcrumbNav).getByText("Root taxonomy");
			expect(currentPageCrumb).toBeInTheDocument();
			expect(currentPageCrumb.tagName).toBe("SPAN");
		});

		it("should render root taxonomy breadcrumb for child taxonomy", () => {
			render(<BorderlineSubstancesSectionPage {...isNotRootProps} />);

			const breadcrumbNav = screen.getByRole("navigation", {
				name: "Breadcrumbs",
			});
			const parentPageCrumb = within(breadcrumbNav).getByText("Root taxonomy");
			expect(parentPageCrumb).toBeInTheDocument();
		});
	});

	describe("body", () => {
		it("should render links to records for taxonomies with one level of children", () => {
			render(<BorderlineSubstancesSectionPage {...isRootSingleLevelProps} />);
			expect(screen.getByRole("link", { name: "Taxonomy 1" })).toHaveAttribute(
				"href",
				"/borderline-substances/taxonomy-1/"
			);
			expect(screen.getByRole("link", { name: "Taxonomy 2" })).toHaveAttribute(
				"href",
				"/borderline-substances/taxonomy-2/"
			);
		});

		it("should render headers for the first level children when there are more than one level of children", () => {
			render(<BorderlineSubstancesSectionPage {...isRootTwoLevelProps} />);
			const heading2 = screen.getAllByRole("heading", {
				level: 2,
			});
			expect(heading2.length).toEqual(2);
			expect(heading2[0]).toHaveTextContent("Taxonomy 1");
		});

		it("should render link back to root taxonomy for borderline substance presentation page", () => {
			render(<BorderlineSubstancesSectionPage {...isNotRootProps} />);
			const link = screen.getAllByRole("link");
			expect(link[4]).toHaveTextContent("View other Root taxonomy");
		});

		it("should flatten all the substances for the section nav", async () => {
			render(<BorderlineSubstancesSectionPage {...isNotRootProps} />);
			const navigation = screen.getByRole("navigation", {
				name: "Navigate to section",
			});

			const findNavigationList = () =>
				within(navigation).findAllByRole("listitem");
			const navigationList = await findNavigationList();
			expect(navigationList.length).toEqual(3);
		});

		it("should render each substance", async () => {
			render(<BorderlineSubstancesSectionPage {...isNotRootProps} />);
			const substance = screen.getAllByRole("heading", { level: 2 });
			expect(substance.length).toEqual(4);
			expect(substance[1]).toHaveTextContent("Fresubin® 1500 Complete");
		});

		it("should render each preparation", async () => {
			render(<BorderlineSubstancesSectionPage {...isNotRootProps} />);
			const preparation = screen.getAllByRole("heading", { level: 3 });
			expect(preparation.length).toEqual(3);
			expect(preparation[0]).toHaveTextContent(
				"Fresubin 1500 Complete liquid Fresenius Kabi Ltd"
			);
		});
	});
});
