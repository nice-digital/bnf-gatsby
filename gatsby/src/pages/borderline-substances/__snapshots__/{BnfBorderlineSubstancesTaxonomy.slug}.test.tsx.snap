// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BorderlineSubstancesSectionPage should match snapshot for graphql query 1`] = `
"
	query ($id: String) {
		bnfBorderlineSubstancesTaxonomy(id: { eq: $id }) {
			slug
			title
			rootTaxonomy {
				slug
				title
				childTaxonomies {
					title
					childTaxonomies {
						title
						slug
					}
					slug
				}
			}
			substances {
				title
				introductionNote
				id
				presentations {
					...FullPresentation
				}
			}
			childTaxonomies {
				title
				substances {
					title
					introductionNote
					id
					presentations {
						...FullPresentation
					}
				}
				childTaxonomies {
					title
					substances {
						title
						introductionNote
						id
						presentations {
							...FullPresentation
						}
					}
				}
			}
		}
	}
"
`;
